### DISCUSSION TITLE
parse syslog date without year

### DISCUSSION THREAD

**Discussion by coredump17** [vectordotdev/vector#19176] [General]
Hi All,

I am new to vector. I am trying to parse network syslogs but cannot use parse_syslog as some logs have an extra field,  not standard format. I can parse my logs with regex but i want to use the timestamp of the actual message, not when it arrived.  The time/date format is "Nov 15 19:45:00" - without the year ;(. 

_$ parse_timestamp!("Nov 15 19:45:00", format: "%h %e %T")
function call error for "parse_timestamp" at (0:55): Invalid timestamp "Nov 15 19:45:00": input is not enough for unique date and time._

testing in vrl, it looks like parse_timestamp requires a year and does not default to 'this year'. Can someone help me overcome this?  Probably easy when you know how ;).

**Comment by polarathene** [comment]
> The time/date format is "Nov 15 19:45:00" - without the year ;(.

That is BSD-Syslog (RFC 3164) formatted [`TIMESTAMP`](https://datatracker.ietf.org/doc/html/rfc3164#section-4.1.2) field, which omits the year and timezone. The properly standardized Syslog V1 format is RFC 5424 and it's [`TIMESTAMP`](https://datatracker.ietf.org/doc/html/rfc5424#section-6.2.3) field uses an RFC 3339 timestamp with additional constraints.

---

> cannot use [`parse_syslog`](https://vector.dev/docs/reference/vrl/functions/#parse_syslog) as some logs have an extra field, not standard format.

> I can parse my logs with regex but i want to use the timestamp of the actual message, not when it arrived

Perhaps it would be better to try detect and extract that extra field that makes it incompatible with `parse_syslog`?

That said [`parse_timestamp`](https://vector.dev/docs/reference/vrl/functions/#parse_timestamp) takes in format that can use [these tokens](https://docs.rs/chrono/latest/chrono/format/strftime/index.html#specifiers).

I am familiar with the format you want for RFC 3164 timestamp due to [this recent syslog PR](https://github.com/vectordotdev/vector/pull/17668#discussion_r1390312110).

Try:

```
parse_timestamp!("Nov 15 19:45:00", format: "%b %e %H:%M:%S")
```

Oh, just realized that your `%h %e %T` format is actually equivalent to that 😅 

I guess this is an issue with `parse_timestamp` support 🤔 

> testing in vrl, it looks like `parse_timestamp` requires a year and does not default to 'this year'. Can someone help me overcome this?

I'm unable to test myself right now but presumably you could just provide/prepend the current year? Perhaps with [`join`](https://vector.dev/docs/reference/vrl/functions/#join)?

**Reply by jszwedko** [reply]
Yeah, the issue here is that there is no year so the timestamp cannot be parsed. I'd second the suggestion of appending the year before parsing.

**Reply by polarathene** [reply]
> I'd second the suggestion of appending the year before parsing.

Is there a way to do that easily? It'd seem you'd have to call `now()` to get the current date and extract the year from that?

A static year would work ok until new years arrives and someone has to manually/automate the date?

If a year is missing using the current year as a fallback seems appropriate? That is the assumption with RFC 3164 (BSD Syslog) timestamps and perhaps elsewhere too.

**Reply by jszwedko** [reply]
You could use `now()` along with `format_timestamp()` to get the current year.

One tricky bit to manage is timestamps that come in close to the new year. If it is January 1st and you see a timestamp for December 31st that is probably a timestamp from the previous year rather than the current one. You may need to compare the month to the current month and choose the year to use accordingly.

I guess, put another way, what would you _like_ to happen? Would you like it to inject the current year when parsing a timestamp that is lacking a year? We could _maybe_ add an an optional parameter to `parse_timestamp` to make that happen.

**Reply by polarathene** [reply]
> We could _maybe_ add an an optional parameter to `parse_timestamp` to make that happen.

That's an interesting scenario with the month concern, but I think if you have an implicit year fallback condition it could probably infer that by the date becoming so far into the future?

At least I'm not sure why someone would expect logs to be processed by Vector that are claiming to be roughly a year into the future?

**Reply by jszwedko** [reply]
That's a good thought, that could be another heuristic to determine the year to be used: is the timestamp far in the future.

**Reply by jszwedko** [reply]
Either way these heuristics are imperfect which I imagine is why RFC5424 syslog messages have a year in the timestamp 😅

**Reply by coredump17** [reply]
It looks like logstash adds the year if one is not provided. Assume that will be taken from now(). Having an option to achieve this would be helpful for this type of timestamp. Not ideal but older juniper devices are built on BSD :)
